autoPyTorch/data/tabular_feature_validator.py

```diff
@@ -481,14 +507,16 @@ def impute_nan_in_categories(self,
             if X[column].isna().any():
                 if column not in self.dict_missing_value_per_col:
                     try:
-                        float(X[column].dropna().values[0])
+                        first_value = X[column].dropna().values[0]
+                        float(first_value)
                         can_cast_as_number = True
                     except Exception:
                         can_cast_as_number = False
                     if can_cast_as_number:
                         # In this case, we expect to have a number as category
                         # it might be string, but its value represent a number
-                        missing_value: Union[str, int] = '-1' if isinstance(X[column].dropna().values[0], str) else -1
+
+                        missing_value: Union[str, int] = '-1' if isinstance(first_value, str) else -1
                     else:
                         missing_value = 'Missing!'
 
```

Comment:
L495--L504 (L523--L532)
dict_missing_value_per_col --> missing_value_symbols
missing_value --> missing_value_symbol (because padded value for missing_value must be a symbol)

Separate the process as a function named `get_unused_category_symbol`.
Plus, this function is easily performed by the following:
```
# Let's assume missing_value is already replaced by missing_value_symbol
if missing_value not in X[column].cat.categories:
    pass
elif isinstance(missing_value, str):
    max_length = max(len(c) for c in X[column].cat.categories)
    missing_value_symbol += '0' * max_length
else:
    # min_value is guaranteed to be negative since there exists -1 in categories
    # and min_value must be smaller than -1. So the symbol is always negative.
    min_value = min(c for c in X[column].cat.categories)
    missing_value_symbol = missing_value + min_value
```

In this case, the time complexity is `O(N)` in the worst case where N is the length of the categories array or the summation of the lengths of categories.
On the other hand, the previous implementation requires `O(N^2)` in the worst case (I do not believe it happens in practice, but the readability is not good in this implementation anyways, so I prefer this implementation).

Reply:
The change in performance is insignificant, however, the code is better.